### PR TITLE
fix(chat): tab close/expand correctness + Tauri-native confirms + live tab rename + MCP arg surfacing

### PR DESCRIPTION
**Pull request by jamesbrink**
## Overview

Eleven commits' worth of bug fixes and small features that surfaced during dogfooding on top of #694 / #696 / #697. Themes: file panel correctness, hotkey ergonomics, chat-tab lifecycle, SCM CI status, and Tauri-webview confirm dialogs. Every commit has its own focused tests (Rust + frontend) — total **+102 test cases** added across the branch.

All checks green:
- 1598 frontend tests · `bunx tsc -b` clean · `bun run lint:css` clean · `bun run lint` 0 errors (13 pre-existing warnings, none in changed files)
- 1048 Rust lib tests · `cargo clippy -p claudette -p claudette-server -p claudette-cli --all-targets --all-features` with `RUSTFLAGS="-Dwarnings"` clean

---

### `327c61f2` `fix(files): keep tracked files visible when ignored trees fill the cap`
**User report:** "after a commit/merge the file explorer only shows ignored files."

**Root cause:** PR #694 dropped `--exclude-standard` from `git ls-files --cached --others -z` so gitignored content (e.g. agent-generated docs) would surface in the Files panel. Side effect: `git ls-files` outputs paths in collated order and `.`-prefixed names sort before `A-Z` < `a-z`, so a worktree with large dot-prefixed ignored dirs (`.codex/`, `.direnv/`, `.logs/`) could exhaust `MAX_FILES = 10,000` before any tracked file was emitted. The frontend re-derives directories from emitted file paths, so the visible tree mirrored the truncated set: ignored dirs only.

**Fix:** Two-pass enumeration in `src-tauri/src/commands/files.rs` —
1. `--cached --others --exclude-standard -z`: tracked + untracked-not-ignored, capped at MAX_FILES. These always win.
2. `--others --ignored --exclude-standard -z`: ignored-only, with a per-top-level-directory cap (`MAX_IGNORED_FILES_PER_TOP_DIR = 200`) so no single noisy ignored tree can dominate.

Refactors the streaming loop into `stream_ls_files` + `record_file_entry` helpers so both passes apply identical filters. New tests pin the regression and the per-bucket fairness contract.

---

### `364f0fcf` `feat(hotkey): rebind Cmd+T to context-aware new tab`
**User report:** "rebind Cmd+T from thinking-mode to context-aware new-tab."

Removes the raw `window.addEventListener("keydown")` listener in `ChatToolbar` / `ComposerToolbar` that toggled thinking mode (and bypassed the keybinding system entirely), plus the `<kbd>⌘T</kbd>` chip-overlay badges. Adds a registered `global.new-tab` HOTKEY_ACTION (`mod+t` cross-platform) that routes by active right-pane view: file tab → triggers FilesPanel inline-create flow at workspace root via a per-workspace nonce; otherwise → creates a fresh chat session and switches to it. `terminal.new-tab` is unaffected (separate scope, dispatched from a different listener).

No DB migration — old binding lived as raw listeners, never persisted. New action appears in **Settings → Keyboard**, rebindable.

---

### `4d2744a5` `fix(scm): add Skipped CI status so merged PRs don't show phantom Running`
**User report:** Skipped CI job was rendered as "Running" on a merged PR.

Both SCM plugins were dropping non-canonical statuses into "pending", which the frontend renders as "Running" — so a merged PR whose only non-success check was `skipped` (GitLab `rules:` excluded) or `SKIPPED`/`NEUTRAL` (GitHub workflow `if:` was false) showed a phantom Running spinner.

End-to-end fix:
- Rust: new `Skipped` variant in `CiCheckStatus` (`src/scm/types.rs`) + serde-roundtrip test
- `plugins/scm-gitlab/init.lua`: `normalize_job_status` maps `skipped` and `manual` → `"skipped"`
- `plugins/scm-github/init.lua`: `normalize_check_status` maps `SKIPPED` and `NEUTRAL` → `"skipped"`; `TIMED_OUT` → `"failure"`. The `statusCheckRollup` aggregations no longer treat SKIPPED/NEUTRAL as breaking "all pass"
- Frontend: extends `CiCheck` status union, label `"Skipped"`, sort priority between cancelled and success, `CircleSlash` icon with muted tone in PrStatusBanner
- `deriveScmCiState` treats skipped-only as success at the aggregate level

---

### `03f3e833` `fix(chat): persist pending attachments across composer remounts`
**User report:** Pasting multiple images into chat then Cmd-Tabbing away loses the attachments.

Root cause: `<ChatPanel>` is conditionally unmounted in `AppLayout.tsx:130-140` whenever the right-pane view changes (file open, diff open, etc.). Component-local `useState` for `pendingAttachments` died with it. The codebase already documents the right pattern at `AppLayout.tsx:96-103` ("Always mount … swap visibility via CSS display:none") for state-bearing panels like Settings and TerminalPanel — ChatPanel was the outlier.

Fix: move attachments to `chatSlice.pendingAttachmentsBySession: Record<string, StoredAttachment[]>`, sibling to the existing `chatDrafts` slot. Composer mirrors every paste/drop/remove/send to the slice; on mount, `hydratePendingFromSlice` rebuilds preview blob URLs from `data_base64` (the underlying Blob is GC'd on the previous mount, so blob URLs intentionally aren't persisted). `removeChatSession` cleans the corresponding slice entry.

8 new tests pin per-session isolation, ordering, single-id removal, full-list replace, and cascading cleanup on session removal.

---

### `c17d9ba8` `fix(chat): restore expand toggle on running tool groups`
**User report:** "When chat messages are grouped and the session is running I should be able to expand them as needed; the chevron used to be there."

PR #696's chronological tool-display split lost two affordances on the grouped path: the chevron decoration and the click handler. The new auto-expand-while-running default was right, but with no toggle the user couldn't drill into a finished group, hide a noisy still-running one, or peek at a group whose status was about to flip.

Restores both: chevron glyph (`⌄` / `›`), `role="button"` + `tabIndex` + `aria-expanded` + `onClick`/`onKeyDown` on the header. State: `userOverride: boolean | null` inside `GroupedToolActivityRows` — `null` follows the running-or-search-match default, a click pins explicit choice. Override survives appended activities because the parent now keys this child by `${first toolUseId}`. Inline display mode unchanged (intentionally flat per #696).

---

### `3d6e0d8b` `feat(chat): surface MCP tool args via a registry + field heuristics`
**User report:** `mcp__postgres__query` shows nothing in the row — should show the SQL.

Pre-fix, `extractToolSummary`'s MCP fallback only checked four field names (`description`/`url`/`query`/`command`). The `sql` field used by every common postgres MCP server fell through silently.

Adds `src/ui/src/components/chat/toolMetadata.ts`, a 3-layer resolver returning `{ summary, lang, fullContent }`:
1. **Exact-match registry** — built-in Claude tools + well-known MCP servers (postgres / sqlite / mysql / bigquery / redshift / playwright / github / filesystem / fetch / memory). Each entry pins the input field carrying the user-facing payload + the Shiki language id.
2. **Tool-name heuristics** — patterns like `mcp__*__query` / `mcp__*__evaluate` / names containing `postgres|mysql|sqlite|redshift|bigquery|snowflake|clickhouse|duckdb` infer SQL/JS/bash. New MCP servers following common conventions get a reasonable summary without manual registry entries.
3. **Field-name heuristics** — first known content field present (sql / code / function / script / command / prompt / query / text / body / title / description / url / path / file_path / name / q / pattern). Falls through to the longest string-valued field, then empty.

15 tests cover all three layers + edge cases (invalid JSON, non-object input, ellipsis truncation, TodoWrite count).

The `lang` and `fullContent` fields are returned but not yet rendered — they scaffold a follow-up: **#698 expand-on-click tool-call details with Shiki syntax highlighting** (assigned, ready to pick up).

---

### `5cb10296` `feat(hotkey): Cmd+W global close-tab + Monaco Cmd+T/Cmd+W bypass`
**Two related user reports:**
- "Cmd+T doesn't create a new file when Monaco editor has focus" — Monaco's default `Cmd+T = Go To Symbol` swallowed the keystroke before the window listener fired.
- "Cmd+W doesn't close chat tabs" — `file-viewer.close-file-tab` was scoped to `file-viewer` only.

DRY architectural fix:
- New `src/ui/src/hotkeys/contextActions.ts` with `executeNewTab`/`executeCloseTab`/`chatCloseConfirmKind` — pure routing functions with dependency-injected services (testable without Tauri).
- `useKeyboardShortcuts` and `MonacoEditor.editor.addCommand` both call into the shared module so Monaco/window-listener never diverge.
- Renamed `file-viewer.close-file-tab` → `global.close-tab` (scope flips file-viewer → global). SQL migration `20260509000540_rename_close_file_tab_keybinding.sql` carries any user-customised binding forward (mirrors `rename_cycle_workspace_keybindings.sql`). Two Rust tests pin the rename direction + the both-keys-present case.

Confirmation rules shared between close-button and Cmd+W via `chatCloseConfirmKind`:
- `running` → existing prompt (preserved)
- `active` → NEW: closing the currently-selected chat asks first
- `last` → NEW: closing the only remaining session asks first
- `none` → close button on a non-active, non-last, non-running tab is silent

22 new tests across `contextActions.test.ts` + `bindings.test.ts` + `useAppStore.fileTree.test.ts` + `db/mod.rs` (migration).

---

### `b79bb2ce` `fix(chat): per-group tool collapse + 3 Codex review fixes + CLI banner wrap`
Bundles five small fixes from a Codex peer review and continued dogfooding:

1. **Per-group tool collapse** (user report: clicking one tool block collapsed all sibling groups in the same turn). Chronologically-split groups all read the same `turn.collapsed`. Adds `collapsedToolGroupsBySession` slice keyed by `${turn.id}:${first toolUseId}`. Single-group turns still flip `turn.collapsed` so persistence-aware code stays accurate; multi-group turns leave it alone.
2. **CLI invocation banner header wrap** (user report: "this header should always fully fit its content"). `.summary` switched from `nowrap + ellipsis` → `white-space: normal` + `overflow-wrap: anywhere` so the chip wraps and grows vertically rather than vanishing.
3. **[Codex P2] FilesPanel + FileViewer nonce mount race** — replaced per-instance number ref with a per-workspace `Map` seeded at 0 so a bump that landed during mounting still gets consumed.
4. **[Codex P2] Search match should override user-collapsed groups** — restructured `isExpanded = queryHasMatch || (userOverride ?? defaultExpanded)` so search hits never land in detached DOM.
5. **[Codex P3] PDF rehydrate fallback** — render the filename+size badge whenever `preview_url` is empty; the renderer no longer breaks PDFs into broken ``s after a remount.

Plus 6 new slice tests and 1 search-regression test.

---

### `c17075b6` `fix(chat): unify tool-group expand state across running→completed + tighten close guards + CLI banner shrinkproof`
Three more bundled user-reported issues:

1. **Tool-group expand state didn't survive turn end.** Live `GroupedToolActivityRows` used local `useState`; completed `TurnSummary` used the slice. Two state sources → expanding a running group silently collapsed when the turn finished. Fix: new `collapsedToolGroupKey(activities)` helper produces `${kind}:${first toolUseId}` (no turn-id prefix) — the same key on both sides of the running→completed boundary because the `toolUseId` is preserved verbatim during `finalizeTurn`.
2. **Bulk-close path lacked active/last guards.** `closeEntries` (Cmd+Shift+W "close all") only confirmed for *running* sessions, so a single keystroke could silently yank the visible chat or wipe out the only-remaining session. Now mirrors the per-tab path's priority order (running → active → last → none) with one prompt per batch.
3. **CLI banner kept collapsing to a thin stripe** (user report: "shrinking and collapsing away"). Two defensive CSS guards: `flex-shrink: 0` on `.banner` so the block reserves vertical space even under sibling overflow pressure; `min-height: 32px` floor on `.header` so the chip is always tappable regardless of zoom.

5 new tests for `collapsedToolGroupKey` + 1 ToolActivitiesSection write-through test.

---

### `80b889dd` `fix(chat): live-update chat tab name from Haiku auto-rename event`
**User report:** "Tab should be renamed after initial prompt like the workspace is — right now it only works if you go to another project and come back."

Root cause: a perfect symmetric miss. The Rust side emitted both `workspace-renamed` and `session-renamed` events from parallel auto-rename background tasks, but the frontend only listened for the workspace half. The session-rename DB write happened correctly; the renamed value only surfaced when `SessionTabs`' mount effect re-fetched the list — which is exactly what a workspace switch triggered.

Adds the missing `session-renamed` listener in `useAgentStream.ts` mirroring the workspace-renamed pattern. New `useAgentStream.sessionRenamed.test.tsx` mounts the hook in happy-dom under a Tauri-event listener mock that captures handlers by name, then drives the captured `session-renamed` handler with payloads. 3 cases: listener IS registered, matching session updates live, unknown id is a safe no-op.

---

### `72e0ce50` `fix(chat): use Tauri's async ask() instead of window.confirm (silent no-op in webview)`
**User report:** "Cmd+W kills running sessions despite them being running or the last tab open — no confirmation appears."

Root cause: `window.confirm()` in Tauri 2's WKWebView (and the WebKit/WebView2 equivalents) does NOT show a native dialog. It returns immediately, so the close path runs unchecked even though the source code has the right `if (!window.confirm(msg)) return` guard. Earlier branch commits added the right *decision logic* (running / active / last → confirm) but kept calling `window.confirm`, so the dialog never surfaced.

**Whole-class** fix: every `window.confirm` in the codebase was a silent no-op in production. Routed through `@tauri-apps/plugin-dialog`'s async `ask()` (already permitted via `dialog:default` in `tauri.conf.json`):
- `hotkeys/contextActions.ts` — `ContextActionDeps.confirm` is now `(message) => Promise<boolean>`. `executeCloseTab` moves the confirm + archive into a single async IIFE so the dialog can be `await`ed. Adds a stale-state recheck after the modal returns — the user could've switched sessions while the dialog was open.
- `components/chat/SessionTabs.tsx` — new `askToClose(message)` helper, used by both `handleArchive` (× button) and `closeEntries` (bulk close).
- `components/terminal/TerminalPanel.tsx` — drive-by; `handleStopAgentTask` had the same bug.

New regression case in `contextActions.test.ts`: "prompts and archives a Running session" — fires Cmd+W on a running session, asserts the confirm dispatches and a "no" answer leaves the session alive. Existing tests converted from `mockReturnValue` → `mockResolvedValue` for the new async contract.

---

## Files changed

**Rust:** `src/scm/types.rs`, `src-tauri/src/commands/files.rs`, `plugins/scm-{github,gitlab}/init.lua`, `src/migrations/mod.rs` + 1 new SQL migration, `src/db/mod.rs` (migration tests)

**Frontend (TS/TSX):** ~25 files across `hotkeys/`, `hooks/`, `stores/slices/`, `components/{chat,file-viewer,files,right-sidebar,terminal}/`, `services/tauri.ts`, `types/{chat,plugin}.ts`, `utils/scmChecks.ts`

**Tests added:** ~12 new test files / suites covering the new code paths

**Locales:** all five locale `chat.json` + `settings.json` files updated for the new keyboard action descriptions and `session_active_confirm_close` / `session_last_confirm_close`

---

## Follow-up

Filed during this branch: [#698 expand-on-click tool-call details with Shiki syntax highlighting](https://github.com/utensils/claudette/issues/698) — the `lang` and `fullContent` fields returned by the new tool registry already scaffold this. Picks up cleanly on top of this PR.

### CODE REVIEW

File: src/ui/src/components/chat/SessionTabs.tsx [hunk 1]
```diff
@@ -29,6 +29,21 @@ import {
...
+async function askToClose(message: string): Promise<boolean> {
+  return ask(message, {
+    title: "Close session",
+    kind: "warning",
+    okLabel: "Close",
+    cancelLabel: "Cancel",
+  });
+}
 import { SessionStatusIcon, type SessionStatusKind } from "../shared/SessionStatusIcon";
 import { DangerousFlagBadge } from "./DangerousFlagBadge";
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/SessionTabs.tsx:35]: Addressed in 135ef912 — moved the `askToClose` helper below the import block (next to the other module-scoped helpers) so the imports form a single contiguous group at the top.

File: src/ui/src/locales/zh-CN/chat.json [hunk 2]
```diff
@@ -88,6 +88,8 @@
   "session_close_diff": "关闭差异视图",
   "session_running_confirm_close": "此会话仍在运行中。停止并关闭 “{{name}}” 吗？",
   "session_running_confirm_close_multi": "{{count}} 个会话仍在运行中。停止并关闭它们吗？",
+  "session_active_confirm_close": "Close the active session \"{{name}}\"?",
+  "session_last_confirm_close": "Close the only remaining session \"{{name}}\"? A fresh session will replace it; the closed one will be archived.",
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/locales/zh-CN/chat.json:93]: Addressed in 135ef912 — translated to:
- `"关闭当前活动的会话 "{{name}}" 吗？"`
- `"关闭仅剩的会话 "{{name}}" 吗？将创建一个新会话替代它；被关闭的会话会被归档。"`
Matches the style of the existing `session_running_confirm_close` translation.

File: src/ui/src/locales/pt-BR/chat.json [hunk 3]
```diff
@@ -88,6 +88,8 @@
   "session_close_diff": "Fechar diff",
   "session_running_confirm_close": "Esta sessão ainda está em execução. Parar e fechar \"{{name}}\"?",
   "session_running_confirm_close_multi": "{{count}} sessões ainda estão em execução. Parar e fechá-las?",
+  "session_active_confirm_close": "Close the active session \"{{name}}\"?",
+  "session_last_confirm_close": "Close the only remaining session \"{{name}}\"? A fresh session will replace it; the closed one will be archived.",
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/locales/pt-BR/chat.json:93]: Addressed in 135ef912 — translated to:
- `"Fechar a sessão ativa "{{name}}"?"`
- `"Fechar a única sessão restante "{{name}}"? Uma nova sessão será criada no lugar dela; a sessão fechada será arquivada."`

File: src/ui/src/locales/ja/chat.json [hunk 4]
```diff
@@ -88,6 +88,8 @@
   "session_close_diff": "差分を閉じる",
   "session_running_confirm_close": "このセッションは実行中です。「{{name}}」を停止して閉じますか？",
   "session_running_confirm_close_multi": "{{count}} 件のセッションが実行中です。停止して閉じますか？",
+  "session_active_confirm_close": "Close the active session \"{{name}}\"?",
+  "session_last_confirm_close": "Close the only remaining session \"{{name}}\"? A fresh session will replace it; the closed one will be archived.",
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/locales/ja/chat.json:93]: Addressed in 135ef912 — translated to:
- `「アクティブなセッション「{{name}}」を閉じますか？」`
- `「最後に残っているセッション「{{name}}」を閉じますか？新しいセッションが代わりに作成され、閉じたセッションはアーカイブされます。」`

File: src/ui/src/locales/es/chat.json [hunk 5]
```diff
@@ -88,6 +88,8 @@
   "session_close_diff": "Cerrar diff",
   "session_running_confirm_close": "Esta sesión sigue en ejecución. ¿Detener y cerrar «{{name}}»?",
   "session_running_confirm_close_multi": "{{count}} sesiones siguen en ejecución. ¿Detener y cerrarlas?",
+  "session_active_confirm_close": "Close the active session \"{{name}}\"?",
+  "session_last_confirm_close": "Close the only remaining session \"{{name}}\"? A fresh session will replace it; the closed one will be archived.",
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/locales/es/chat.json:93]: Addressed in 135ef912 — translated to:
- `"¿Cerrar la sesión activa «{{name}}»?"`
- `"¿Cerrar la única sesión restante «{{name}}»? Se creará una nueva en su lugar; la sesión cerrada se archivará."`
Matches the `«»` quoting convention used in the existing es `session_running_confirm_close`.

File: src/ui/src/components/chat/toolMetadata.ts [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+ */
+
+export interface ToolDisplayMeta {
+  /** Input field whose value is the primary content to surface as
+   *  the tool-call summary. Must be a string-valued field. */
+  contentField: string;
+  /** Optional Shiki language id for highlighting the field's value
+   *  when rendered as a code block. `null` / absent renders plain. */
+  contentLang?: string | null;
+}
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/toolMetadata.ts:46]: Addressed in 135ef912 — updated `ToolDisplayMeta.contentField`'s doc to describe the array + object handling that `displayFromField` already implements (e.g. `todos` → count, other arrays → joined string, objects → JSON-stringified). Kept the type as `string` (the field name itself, not the value type) since that's accurate.

File: src/ui/src/hotkeys/actions.ts [hunk 7]
```diff
@@ -313,10 +339,23 @@ export const HOTKEY_ACTIONS = [
...
+    //  - Diff active → closes the diff tab.
+    //  - Chat active → archives the active chat session, gated by the
+    //    shared confirm rules in `chatCloseConfirmMessage` (running
+    //    sessions, the active session, and the last remaining session
+    //    all prompt before close).
+    //
+    // Replaces the prior `file-viewer.close-file-tab` action; the
+    // `20260508T_rename_close_file_tab_keybinding.sql` migration carries
+    // any user-customised binding forward.
+    id: "global.close-tab",
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/hotkeys/actions.ts:355]: Addressed in 135ef912 — fixed the comment to reference the actual migration filename `20260509000540_rename_close_file_tab_keybinding.sql`. The bogus `20260508T_…` placeholder was a draft id I forgot to update before settling on the real timestamp.

File: plugins/scm-gitlab/init.lua [hunk 8]
```diff
@@ -90,12 +90,26 @@ function M.merge_pull_request(args)
...
 -- Normalize GitLab job status to canonical CiCheckStatus values.
--- GitLab returns: created, pending, running, success, failed, canceled, skipped, manual.
+-- GitLab returns: created, pending, running, success, failed, canceled,
+-- skipped, manual. Map them onto the four-canonical statuses Rust
+-- consumes:
+--   success                 → "success"
+--   failed                  → "failure"
+--   canceled                → "cancelled"  (note GitLab's "canceled" — single l)
+--   skipped, manual         → "skipped"   (didn't run by design / awaiting manual trigger)
+--   created, pending, running → "pending" (in-flight or queued)
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply plugins/scm-gitlab/init.lua:100]: Addressed in 135ef912 — updated the comment to "five canonical CiCheckStatus variants". The wording was correct pre-PR (four variants: Pending / Success / Failure / Cancelled) but this PR adds the `Skipped` variant, so the count grew to five.
